Convert Istio dashboards to Kibana Lens

gsantoro opened the pull request:
## What does this PR do?

Convert the Istio dashboards to Lens

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

- Relates #4678

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

constanca-m commented:
Seems ok. For the jenkins error, update the kibana requirement [here](https://github.com/elastic/integrations/blob/7d25b6e28bc538206b6e2dad2f67bb593d39a7cc/packages/istio/manifest.yml#L14) to `8.6.0`, or change the version [here](https://github.com/elastic/integrations/blob/7d25b6e28bc538206b6e2dad2f67bb593d39a7cc/packages/istio/kibana/dashboard/istio-f0f863b0-5941-11ed-bfb8-bbfe5b338339.json#L2724) and [here](https://github.com/elastic/integrations/blob/7d25b6e28bc538206b6e2dad2f67bb593d39a7cc/packages/istio/kibana/dashboard/istio-f0f863b0-5941-11ed-bfb8-bbfe5b338339.json#L2720) for the dashboard (`8.4.0` if possible).

gizas commented on packages/istio/kibana/dashboard/istio-f0f863b0-5941-11ed-bfb8-bbfe5b338339.json:
```diff
@@ -205,111 +354,197 @@
+                                                    "dataType": "number",
+                                                    "isBucketed": false,
+                                                    "label": " ",
+                                                    "operationType": "max",
```
Should this be max?

gsantoro replied:
About this metric. I don't even know what it's doing but it is not one of those ones reported by Stephen to convert from Max to Sum. Before I change that, I'll have a chat with Stephen.

gizas commented on packages/istio/kibana/dashboard/istio-f0f863b0-5941-11ed-bfb8-bbfe5b338339.json:
```diff
@@ -826,978 +1418,1454 @@
-                "title": "Galley Validation Failed",
-                "type": "visualization",
-                "version": "8.4.0"
+                "title": "Galley Validation Failed by Label",
```
Just tempted to ask if we need Galley in general. Do you think it gives value? Can we remove it?

gsantoro replied:
I have no idea what it is yet. @ChrsMark do you know?

ChrsMark replied:
Istiod still reports this metric, right? If so, why to remove it?
The definition of this metric seems to be `Count of validations failed by the configuration validation webhook`. The metric is reported as `galley_validation_failed{group="networking.istio.io",reason="invalid_resource",resource="gateways",version="v1alpha3"} 1`.

So I would say if it's still provided, it's a valuable indicator that Istio provides and some other tools use it to create alerts. See an example [here](https://www.datadoghq.com/blog/how-to-monitor-istiod/#metrics-to-alert-on-galley_validation_failed-galley_validation_passed). So I wouldn't remove it if there is room for it in the dashboard but it's up to you folks to make the last call.

gsantoro replied:
I'll have a look at that metric and I'll write here about it.

gizas commented:
We can keep adding short info for how we test and also screenshots to provide evidence for your results. 
Dashboard stories with screenshot evidence should be a must. 

Also in general the description text like `Metrics from Istiod` does not give much value. Either remove it or enhance it with valuable information on how the users should read the dashboards .

Some more ideas like grouping Pilot XDs together or Time related visualisation can also be handy. Especially in overview I am getting lost to read so many information.

Also @constanca-m used markdown to group specific visualisations. You can also group by this your visualisations like here https://github.com/elastic/beats/pull/34161

Overall great improvement Giuseppe!

gsantoro commented:
> We can keep adding short info for how we test and also screenshots to provide evidence for your results. Dashboard stories with screenshot evidence should be a must.

I agree. I opened this PR in `draft` mostly to see what was broken with the Jenkins build before adding screenshots and instructions to test.

> Also in general the description text like `Metrics from Istiod` does not give much value. Either remove it or enhance it with valuable information on how the users should read the dashboards .

I am not sure what to put in there. I just wanted to convey that those metrics in `Overview` are from Istiod compared to the ones in `Traffic` which have a description `Metrics from Istio Sidecar Proxies`. I am open to suggestions.

> Some more ideas like grouping Pilot XDs together or Time related visualisation can also be handy. Especially in overview I am getting lost to read so many information.
> 
> Also @constanca-m used markdown to group specific visualisations. You can also group by this your visualisations like here [elastic/beats#34161](https://github.com/elastic/beats/pull/34161)

Looks good. I'll have a look into those. Thanks

> Overall great improvement Giuseppe!

gsantoro commented:
@constanca-m and @gizas 

> Also @constanca-m used markdown to group specific visualisations. You can also group by this your visualisations like here https://github.com/elastic/beats/pull/34161

It looks great that piece of markdown. Unfortunately it's a TSVB panel and the styling with custom CSS is not available in Lens. Given that the topic of this PR is to move away from TSVB I am not sure that I should add a TSVB panel for styling some markdown.

![Screenshot 2023-02-15 at 17 11 22](https://user-images.githubusercontent.com/2378291/219102532-1e6c2a04-e712-4c0b-91e3-08cc8f306589.png)

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`1/1`) | :green_heart: 
Files | 100.0% (`3/3`) | :green_heart: 
Classes | 100.0% (`3/3`) | :green_heart: 
Methods | 90.0% (`27/30`) | :+1: 65.0
Lines | 97.472% (`347/356`) | :-1: -2.528
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

gizas commented:
- Ok markdown in TSVB is not something we must do it. Just group then the related visualisations 
- For suggestions for text: You can even put links to Istio pages that describe the components you visualise for reference.
- Keep Galley 
- Have a look for the max comment https://github.com/elastic/integrations/pull/5268#discussion_r1107282884

gsantoro commented:
I would like to start a discussion about the changes applied to migrate from TSVB to Lens. I'll highlight every change in a separate comment so that we easily reference them and comment them.

1. in Lens you can't see the last value for a metric in the Legend. The current value is displayed as usual when you over the metric.

In TSVB
<img width="833" alt="Screenshot 2023-02-16 at 10 24 51" src="https://user-images.githubusercontent.com/2378291/219339475-e33db898-8971-4cb3-bde9-4c9156eff1c3.png">

in Lens
![Screenshot 2023-02-16 at 10 24 57](https://user-images.githubusercontent.com/2378291/219339530-9e563104-73c1-4eac-a15d-41b28ab396f7.png)

gsantoro commented:
2. Requests per source/destination used to use the Max function instead of Sum function. . Now instead, we sum the number of requests per source/destination and we calculate the rate in seconds as well. They have migrated to Lens as well

old implementation in TSVB with Max function
<img width="1668" alt="Screenshot 2023-02-16 at 13 24 35" src="https://user-images.githubusercontent.com/2378291/219377038-10d09b8f-d00c-4889-a042-b7d1507388a8.png">

new implementation in Lens with Sum function
![Screenshot 2023-02-16 at 13 24 51](https://user-images.githubusercontent.com/2378291/219377289-18322c39-d6da-41f7-ac38-457a11557c56.png)

gsantoro commented:
4. Similar discussion as point 3 for Duration of requests by source/destination

in TSVB
<img width="1662" alt="Screenshot 2023-02-16 at 13 31 12" src="https://user-images.githubusercontent.com/2378291/219378288-58a49c9e-a149-4950-8698-fa0712318004.png">

in Lens
![Screenshot 2023-02-16 at 13 30 39](https://user-images.githubusercontent.com/2378291/219378363-be77bd8f-1714-4b73-9f8d-369360a393be.png)

gsantoro commented:
6. Multiple percentiles in the same visualization

in TSVB (notice the last value after the metric is a bit confusing here
<img width="663" alt="Screenshot 2023-02-16 at 13 36 37" src="https://user-images.githubusercontent.com/2378291/219379564-8c1f2530-2489-44a7-9ad2-30b01c76696b.png">


In Lens (percentiles are indicated with `(25)` for 25th percentile, ...
![Screenshot 2023-02-16 at 13 36 47](https://user-images.githubusercontent.com/2378291/219379678-4132a82e-eaa8-4ee5-9c78-64c6a79e3386.png)

gsantoro commented:
7. The final look of both dashboards Overview and Traffic

in TSVB
<img width="1684" alt="Screenshot 2023-02-16 at 13 40 56" src="https://user-images.githubusercontent.com/2378291/219380531-067dbde6-f325-4435-b19d-234372e9baba.png">

<img width="1686" alt="Screenshot 2023-02-16 at 13 40 46" src="https://user-images.githubusercontent.com/2378291/219380522-70cec16c-1022-45c0-ba79-a3e49d89ff21.png">



in Lens
<img width="1685" alt="Screenshot 2023-02-16 at 13 40 21" src="https://user-images.githubusercontent.com/2378291/219380613-98616d22-e599-4db0-ba0b-5c1cafed68c2.png">
<img width="1687" alt="Screenshot 2023-02-16 at 13 39 59" src="https://user-images.githubusercontent.com/2378291/219380606-9c5b88de-7c08-43e0-a7da-87f93382c726.png">

gsantoro commented:
Hey @bvader, 
do you mind having a look at this PR for the changes that we discussed on our last meeting?

gsantoro commented:
8. I recently saw that Kubernetes integration comes with a new way to implement Gauges (like in the below screenshot). 

![Screenshot 2023-02-16 at 15 16 19](https://user-images.githubusercontent.com/2378291/219407854-181e8c5f-23f7-4fd9-8756-ddb779359ac7.png)

The problem with that is it's coming from a Gauge TSVB visualization.

Since there was lot of focus on moving the Istio integration entirely to Lens, I am now wondering how much of hour current dashboards still rely on TSVB given that I have seen at least
- markdown for category of dashboards
- Gauges for percentage usage
- (there might be more I am not aware of yet)

gizas commented:
At the moment all above changes LGTM.
Two minor:
1. Foot Labels like `@timestamp per hour` can be removed? Or to be included in the header of visualisation? 
2. Numeric Metrics in 5: Can you change the actual number to be in the center and big? Also Capitialise headers and make them onliners?

gsantoro commented:
> At the moment all above changes LGTM. Two minor:
thanks for the feedback

> 
> 1. Foot Labels like `@timestamp per hour` can be removed? Or to be included in the header of visualisation?

About the foot label. I can either remove the label entirely or customize the `@timestamp` part. The second part of that label actually change based on the timeframe. So that part might be `per hour` or `per seconds`. I can customize that part other than fixing the time frame or remove that part entirely.

this is how you remove the `@timestamp` part of that label

![Screenshot 2023-02-21 at 14 43 46](https://user-images.githubusercontent.com/2378291/220376978-61cd0a77-8606-4056-a254-959f154aa038.png)

and this is how it looks with no X axis label.

![Screenshot 2023-02-21 at 14 45 00](https://user-images.githubusercontent.com/2378291/220376983-85fba1da-5355-4d41-a835-8d7c3b50da67.png)

What do you suggest?


> 2. Numeric Metrics in 5: Can you change the actual number to be in the center and big? Also Capitialise headers and make them onliners?

I can't modify any of the visual aspect of this Metric visualization, meaning either the aspect of the number or the headers. TSVB has lot more customization capabilities (via CSS) than Lens. The only alternative is to have a single big percentage number but we will be getting rid of the color that indicate a possible range of values.

constanca-m commented:
> It looks great that piece of markdown. Unfortunately it's a TSVB panel and the styling with custom CSS is not available in Lens. Given that the topic of this PR is to move away from TSVB I am not sure that I should add a TSVB panel for styling some markdown.

This is unfortunately true. There are other things that Lens can't do yet, so I wouldn't call having this text as a problem. Example of another thing: summing the last value of some metric that breaks down in a label.

constanca-m commented:
> Foot Labels like @timestamp per hour can be removed? Or to be included in the header of visualisation?

If it were me, I would try to remove the axis labels, and try to be clear with the title and the legend (if it exists). And there is no reason to add the unit type (example, sec, %) in the title, if we can write the axis value with a suffix - like 4 sec. @gsantoro

constanca-m commented:
> let the user guess the timeframe by moving the cursor through the values ... This is how it looks with no X axis and 1h fixed timeframe

The visualization is very understandable, I don't see a problem in not having a suffix in the x values.

gsantoro commented:
Here it is the updated look of the dashboard. Changes are:
- inline all pilot Numeric metrics
- removed X axis from all visualization
- set the default Date Histogram resolution to 1h. Just to be clear. If you have more than a couple of days of time frame, the dashboard use a 3h bucket size instead of 1h. This bucket size changes automatically by default based on the timeframe. Fixing it to 1h might have performance implication if you look at more 1w time frame. I hope everyone is ok with this tradeoff. 

![Uploading Screenshot 2023-02-21 at 17.30.22.png…]()

constanca-m commented:
The link to the screenshot just leads me to the description of the PR @gsantoro

gizas commented:
> * set the default Date Histogram resolution to 1h. Just to be clear. If you have more than a couple of days of time frame, the dashboard use a 3h bucket size instead of 1h. This bucket size changes automatically by default based on the timeframe. Fixing it to 1h might have performance implication if you look at more 1w time frame. I hope everyone is ok with this tradeoff.

Such kind f information can go to a small markdown if you think the user should know. Sth like `Default size of bucket is 1h, you can anticipate performance degradation if large number is set` 

Overall LGTM

gsantoro commented:
@constanca-m sorry about that. I fixed the screenshot now.

@gizas I am personally not a huge fan of adding a markdown text to explain the decision behind a choice that might affect the performance of a dashboard. I would have preferred to keep the time frame dynamic like we do elsewhere and then do not required the markdown. Nevertheless, I am happy to go with this solution if you think it is better.

gizas commented:
Sure no strong opinion on this, just an alternative if you wanted to consider.

gsantoro commented:
I got some feedback from Stephen Brown that the Istiod Dashboards with 3 visualization per row are a `a bit cramped even on a fairly wide screen`. 

This is how it looks with 2 visualization per row instead.

<img width="1429" alt="Screenshot 2023-02-22 at 11 18 00" src="https://user-images.githubusercontent.com/2378291/220605236-5186bb1b-19ca-44d9-b355-4fcbd856e050.png">

I agree with him that it looks better now. So I'll keep it like this unless someone object here.

About the time frame being fixed or not from a previous discussion. This is really what I would like to do:
- keep the time frame dynamic like per default settings so that it doesn't create buckets that are too small or too big, it doesn't impact the performance of the dashboard and we don't have to add the extra bit of markdown
- I will still remove the X axis label `@timestamp per 3h`
- I'll leave to the user to notice the time frame resolution by just moving the mouse over a couple of sample point from each visualization.

Since we don't have strong guidelines on the previous topics, this is what I think it's sensible but feel free to point out to a better way to do this or some documentation.

gsantoro commented:
A couple of small changes:
- @bvader made me notice that we have both "Requests per source/destination" and "Requests / sec per source/destination". So I remove the first 2 in favour of just leaving the last 2.
- I have removed the X labels from all dashboards (this was already agreed but I had some still there)
- I have added the dashboard pre-filters according to https://github.com/elastic/integrations/issues/4978
- I have tested everything with some real Istio data

<img width="1429" alt="Screenshot 2023-03-06 at 18 03 48" src="https://user-images.githubusercontent.com/2378291/223193972-6ef6757c-f8c2-405c-a2d8-4321721648d2.png">
<img width="1432" alt="Screenshot 2023-03-06 at 18 04 01" src="https://user-images.githubusercontent.com/2378291/223193980-e7b67086-2d6e-43eb-a14e-c1b35960873b.png">

ChrsMark reviewed (approved):
LGTM!

elasticmachine commented:
Package istio - 0.2.4 containing this change is available at https://epr.elastic.co/search?package=istio
